### ISSUE TITLE
Question: Is it able to use lightningcss as a parser from node?

### ISSUE THREAD

**Issue by so1ve** [parcel-bundler/lightningcss#440]

**Comment by devongovett** [comment]
This is not currently possible mainly for performance reasons. If you're looking to transform the AST in some way, the [custom transform](https://lightningcss.dev/transforms.html) API can be used. This is faster because we only need to serialize the portions of the AST that you want to modify rather than the whole thing. What are you looking to accomplish?

**Comment by so1ve** [comment]
> This is not currently possible mainly for performance reasons. If you're looking to transform the AST in some way, the [custom transform](https://lightningcss.dev/transforms.html) API can be used. This is faster because we only need to serialize the portions of the AST that you want to modify rather than the whole thing. What are you looking to accomplish?

Well, I want to use lightningcss as a css parser to support linting css in eslint. PostCSS works well, but it is too slow for me :(

**Comment by jacobrask** [comment]
> > This is not currently possible mainly for performance reasons. If you're looking to transform the AST in some way, the [custom transform](https://lightningcss.dev/transforms.html) API can be used. This is faster because we only need to serialize the portions of the AST that you want to modify rather than the whole thing. What are you looking to accomplish?
> 
> 
> 
> Well, I want to use lightningcss as a css parser to support linting css in eslint. PostCSS works well, but it is too slow for me :(

I'm interested in something similar, but for me PostCSS/Stylelint isn't good enough regardless of speed since it's not aware of the different value types in the same way as Lightning.

I want to ensure that only acceptable values for color, spacing, etc are used.

For each lint rule, you'd know which visitors you need, so maybe in fact the current transform API is good enough. If autofixing is enabled you could use the transform as well.

**Comment by so1ve** [comment]
> I'm interested in something similar, but for me PostCSS/Stylelint isn't good enough regardless of speed since it's not aware of the different value types in the same way as Lightning.
> 
> I want to ensure that only acceptable values for color, spacing, etc are used.
> 
> For each lint rule, you'd know which visitors you need, so maybe in fact the current transform API is good enough. If autofixing is enabled you could use the transform as well.

Yup! For me, the biggest problem is that postcss has only several node types (6?). That means it's hard to make custom eslint rules because we cannot determine the context easily. It does not have a good tokenizer either. CSSTree seems to be a good choice, but it is slow too.

**Comment by jacobrask** [comment]
I forked Stylelint to try out using LightningCSS as an optional parser. If I understand the Stylelint/PostCSS architecture correctly it will not do an all parsing until a rule is added with node walkers. That means that you only/mostly pay for the rules you add and could theoretically gradually move over to rules supported by LightningCSS instead.

I made each rule add visitors which are composed and LightningCSS is then ran once. I then pass the output of LightningCSS to PostCSS and run the rest of the Stylelint rules.

What's working:

- Reporting rule violations found with LightningCSS as Stylelint warnings.

What's not working:

- Line numbers/columns
- Fixing. LightningCSS transform will change the original source even for nodes you don't want to touch.
